src/openstack_cloud/openstack_manager.py

```diff
@@ -545,178 +550,271 @@ def __init__(
+            1. SSH fails - could be a transient network error.
+            2. The Runner.* processes do not exist. Mark healthy for now to gather data. This is
+                subject to change to unhealthy once enough data has been gathered.
+
+        Args:
+            conn: The Openstack connection instance.
+            server_name: The name of the OpenStack server to health check.
+            startup: Check only whether the startup is successful.
+
+        Returns:
+            Whether the instance is healthy.
+        """
+        server: Server | None = conn.get_server(name_or_id=server_name)
+        if not server:
+            return False
+        if server.status == (_INSTANCE_STATUS_SHUTOFF, _INSTANCE_STATUS_ERROR):
+            return False
+        if server.status not in (_INSTANCE_STATUS_ACTIVE, _INSTANCE_STATUS_BUILDING):
+            return False
+        return OpenstackRunnerManager._ssh_health_check(
+            conn=conn, server_name=server_name, startup=startup
+        )
 
     @staticmethod
-    def _ssh_health_check(server: Server) -> bool:
+    def _ssh_health_check(conn: OpenstackConnection, server_name: str, startup: bool) -> bool:
         """Use SSH to check whether runner application is running.
 
+        A healthy runner is defined as:
+            1. SSH connection can be established.
+            2. Runner.Worker exists (running a job).
+            3. Runner.Listener exists (waiting for job).
+
         Args:
-            server: The openstack server instance to check connections.
+            conn: The Openstack connection instance.
+            server_name: The openstack server instance to check connections.
+            startup: Check only whether the startup is successful.
 
         Returns:
             Whether the runner application is running.
         """
-        addresses = []
-        for ssh_conn in OpenstackRunnerManager._get_ssh_connections(server=server):
-            addresses.append(ssh_conn.host)
+        try:
+            ssh_conn = OpenstackRunnerManager._get_ssh_connection(
+                conn=conn, server_name=server_name
+            )
+        except _SSHError as exc:
+            logger.error("[ALERT]: Unable to SSH to server: %s, reason: %s", server_name, str(exc))
+            return True
+
+        result: invoke.runners.Result = ssh_conn.run("ps aux", warn=True)
+        logger.debug("Output of `ps aux` on %s stderr: %s", server_name, result.stderr)
+        logger.debug("Output of `ps aux` on %s stdout: %s", server_name, result.stdout)
+        if not result.ok or RUNNER_STARTUP_PROCESS not in result.stdout:
+            logger.warning("List all process command failed on %s ", server_name)
+            return False
+        logger.info("Runner process found to be healthy on %s", server_name)
+        if startup:
+            return True
+
+        if RUNNER_WORKER_PROCESS in result.stdout or RUNNER_LISTENER_PROCESS in result.stdout:
```

Comment:
Is it necessary to check for both, or is it not enough to check for the latter?

Reply:
Only the latter could be checked - but this has been included to match the business logic that was discussed. Later on, this constant can help determine whether the runner is busy or not :)